feat: home studio search filters

johnvente opened the pull request:
### Description

This PR adds a search input and two dropdown menus to filter courses in studio home:

  -  first dropdown menu: Allows us to filter by "All courses," "Active," or "Archived" courses.
  -  second dropdown menus: Allow for ascending/descending course names and the newest/oldest courses.
    
  
**NOTE:** this applies only for "courses" tab

This Uses `ENABLE_HOME_PAGE_COURSE_API_V2` in the environment
  
### DEMO



https://github.com/openedx/frontend-app-course-authoring/assets/134975835/13c02b94-58dc-4929-bcc9-c443d197730b

<img width="786" alt="Screen Shot 2024-04-04 at 8 17 11 AM" src="https://github.com/openedx/frontend-app-course-authoring/assets/134975835/43ed4d33-f61b-44b0-a90a-f544fb92b0a4">


  <img width="336" alt="Screen Shot 2024-02-19 at 8 25 54 PM" src="https://github.com/openedx/frontend-app-course-authoring/assets/134975835/80f2e1af-8385-41ec-b3f3-92af5b60b1d5">
  
<img width="351" alt="Screen Shot 2024-02-19 at 8 26 13 PM" src="https://github.com/openedx/frontend-app-course-authoring/assets/134975835/4b015a60-1eb7-44ad-991e-c0c249a1cc57">


### Supporting information
To review a little bit more about this implementation, please check this:
https://openedx.atlassian.net/wiki/spaces/OEPM/pages/4032856101/Studio+Home+incremental+upgrades+-+product+approach

### How to test it
1. Set up your edx-platform
    ```
     cd path-to-your-edx-platform
     git fetch origin pull/34175/head:MJG/course-home-api-v2-pagination
     git checkout MJG/course-home-api-v2-pagination
    ```
2. Check it out this branch inside frontend-app-course-authoring
    ```
     cd path-to-frontend-app-course-authoring
     git fetch origin pull/846/head:jv/feat-home-studio-filters
     git checkout jv/feat-home-studio-filters
     nvm use && npm install
     npm start
    ```

openedx-webhooks commented:
Thanks for the pull request, @johnvente! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

felipemontoya commented:
I tested this PR when reviewing https://github.com/openedx/edx-platform/pull/34173. It worked great, but I have a minor wording suggestion:

When the sum of filter+search returns an empty list, the message looks like an error happened.
![image](https://github.com/openedx/frontend-app-course-authoring/assets/2816902/c538c315-a354-46b9-9101-f44b35462cd9)
I would only suggest that we make it less error like and just say that "There are no courses with the current filters. --Clear filters"

mariajgrimaldi commented:
I was testing this out and found that when a user doesn't have course creator permissions, this is what is shown:
![image](https://github.com/openedx/frontend-app-course-authoring/assets/64440265/c656d89c-5c72-4664-ba6a-215640aa0a59)
This also happens when no courses are listed for the current user. I'm guessing that's the number of pages. Should we remove it?

johnvente commented:
Note for reviewer: this PR will use `ENABLE_HOME_PAGE_COURSE_API_V2` feature flag when [this one](https://github.com/openedx/frontend-app-course-authoring/pull/825) is merged

mariajgrimaldi reviewed (commented):
I'm guessing all the changes made in https://github.com/openedx/frontend-app-course-authoring/pull/825 so I'll re-review once that's done. However, I still tested the filters locally, and they worked as described in the cover letter. Thank you.

johnvente commented:
I won't be able to review this in a week so I will check any changes 1 April

jmakowski1123 commented:
This was one of the first projects we put through both product review and design first, before implementing, so this already had product review! Original spec [here.](https://openedx.atlassian.net/wiki/spaces/OEPM/pages/4032856101/Studio+Home+incremental+upgrades+-+product+approach)

This looks great, very exciting to see in action!

One very nit-picky thing. When a user types in the search field, do they have to hit enter in order to display the results, or do the results display in real time as they are typing? I'm asking because the in-course search is being implemented now too, and I believe the behavior there is that the results being to display immediately as the user types. It'd be nice to have consistent behavior across both search experiences. Pinging @bradenmacdonald here for more input on the in-course side.

johnvente commented:
> This was one of the first projects we put through both product review and design first, before implementing, so this already had product review and is good to go! Original spec [here.](https://openedx.atlassian.net/wiki/spaces/OEPM/pages/4032856101/Studio+Home+incremental+upgrades+-+product+approach)
> 
> This looks great, very exciting to see in action!
> 
> One very nit-picky thing, @johnvente . When a user types in the search field, do they have to hit enter in order to display the results, or do the results display in real time as they are typing? I'm asking because the in-course search is being implemented now too, and I believe the behavior there is that the results being to display immediately as the user types. It'd be nice to have consistent behavior across both search experiences. Pinging @bradenmacdonald here for more input on the in-course side.

Hi @jmakowski1123 while the user types in the search filed the results display while the user is typing doesn't need to hit enter to see them

johnvente commented:
Hi @jesperhodge! this is like the second part of [pagination](https://github.com/openedx/frontend-app-course-authoring/pull/825) that you reviewed and helped me, could you take a look over here please? this has the changes you recommend me in pagination meantime I will be checking the workflows. Thanks a lot

jesperhodge commented:
Hi @johnvente thanks! I have it on my to-do list and will get to it later today.
FYI I forgot that the V2 feature flag should default to false, so I changed that with https://github.com/openedx/frontend-app-course-authoring/pull/934. Just so you're not confused that it's suddenly changed.

johnvente commented:
Hi @jesperhodge don't worry. Do know why this has changed to false as default? as far I know this feature would be in redwood in backend feature there was a discussion [about that]( https://github.com/openedx/edx-platform/pull/34173#discussion_r1526835238)

felipemontoya commented:
Now it looks like the feature is marked for sumac instead of redwood so I think we can move on leaving this with the default to false and turn it on after the cut. Thank you Jesper and John.

jesperhodge commented:
Hi @johnvente , I still need to review the code but I've tested the feature locally and I'll already share my feedback from the testing side. The search and filters are great! Thanks for the awesome work. There's just a couple of details I wanted to bring up:

- I found a bug where if you type in a search query and then not click x, but remove it with backspace again, often it does not go back to the default and you have no way of clearing the query other than refreshing the page. See my screen recording.
- The UI jumps around a lot while searching. The layout shifts horizontally a little bit, and vertically a lot. Any time I type in a character, every box vanishes, the bottom bar jumps into the screen, and a loading circle appears for a second, before jumping back to results. I discussed that with my product manager @cablaa77 and we'd love to see a bit of a UX improvement here. Now I understand there are a couple of solutions that also have drawbacks. For example, if you do not blend in the loading spinner for a moment, there's not a large shift, and you can just wait until data is loaded and then replace search results with the new ones. But the drawback there then is that the user may experience a delay in response where nothing happens, so that also can be problematic.
Can we find a way to improve this layout shifting to where it feels good to users? It could also be helpful to see if for example @felipemontoya or @jmakowski1123 are able to support to decide on a solution that works well for the user UX-wise, since our own designer is on holiday now.

Other than those points, the feature feels great and I'm excited to get it live!

jesperhodge commented:
https://github.com/openedx/frontend-app-course-authoring/assets/19345795/c84548a8-b2a1-4a17-9194-c42bd5d68cfd

screen recording related to my last comment

felipemontoya commented:
@johnvente my thoughts on this.

- the `x` bug needs to be fixed
- for the shifting layout, could we leave a spinner in the search box (left of the X for example) to show the user that some action is underway, but only replace the list after the results come back

johnvente commented:
Hi @jesperhodge! Thanks a lot for taking a look over here, I will be checking the bug that you found

@felipemontoya we can do that for the shifting layout inside the search box but the main problem is that [SearchField](https://paragon-openedx.netlify.app/components/searchfield/) component doesn't have something to add a new icon (the spinner in this case) but I can do it, even though the X would be clickable and that would be triki. should the "x"  be able to be clickeable when the user is typing or should I hide it?

jesperhodge commented:
> @johnvente my thoughts on this.
> 
> * the `x` bug needs to be fixed
> * for the shifting layout, could we leave a spinner in the search box (left of the X for example) to show the user that some action is underway, but only replace the list after the results come back?

@felipemontoya I love that idea.

johnvente commented on src/studio-home/tabs-section/courses-tab/courses-filters/index.jsx:
```diff
@@ -0,0 +1,142 @@
+  const searchValueDebounced = useDebounce(inputSearchValue);
+
+  const handleSearchCourses = (value) => {
+    setInputSearchValue(value);
```
This is a really great idea, I will implemented it with a use `callback` instead of having that `useEffect`

johnvente commented:
> * I found a bug where if you type in a search query and then not click x, but remove it with backspace again, often it does not go back to the default and you have no way of clearing the query other than refreshing the page. See my screen recording.

Hi @jesperhodge I added the fix for this bug, and changed the debounced flow for your [recommendation](https://github.com/openedx/frontend-app-course-authoring/pull/846#discussion_r1554144223) here


> * for the shifting layout, could we leave a spinner in the search box (left of the X for example) to show the user that some action is underway, but only replace the list after the results come back?
@felipemontoya I added a spinner of the left of the X and I replace the list when it has finished to bring the data


Sorry for the delay, after to day I won't be able to check this pr but please let know to @felipemontoya @mariajgrimaldi any 
any additional change

mariajgrimaldi commented:
This issue is still happening to me: 

> I found a bug where if you type in a search query and then not click x, but remove it with backspace again, often it does not go back to the default and you have no way of clearing the query other than refreshing the page. See my screen recording.

What about you folks?

mariajgrimaldi commented:
I'm still facing this issue: 

> I found a bug where if you type in a search query and then not click x, but remove it with backspace again, often it does not go back to the default and you have no way of clearing the query other than refreshing the page. See my screen recording.

I'm working on fixing it.

mariajgrimaldi commented:
> I'm still facing this issue:
> 
>     I found a bug where if you type in a search query and then not click x, but remove it with backspace again, often it does not go back to the default and you have no way of clearing the query other than refreshing the page. See my screen recording.
> 
> I'm working on fixing it.

I believe I solved this. Now, the coverage upload is still failing.

mariajgrimaldi commented:
@jesperhodge: thank you for running the coverage upload manually, but I think it's also happening in other PRs: 
https://github.com/openedx/frontend-app-course-authoring/actions/runs/8604487493/job/23601521237#step:6:38
https://github.com/openedx/frontend-app-course-authoring/actions/runs/8617944055/job/23619770035#step:6:38
I'm not sure if it's a 3rd party failure or related to the changes themselves.

jesperhodge commented:
Thanks, @mariajgrimaldi . Yeah usually the coverage upload works after retrying, but it seems broken. From my side, I reviewed and everything looks and works great. I'm waiting for a confirmation from @cablaa77 for the UX but I don't expect any objections at all, then I'll approve.
I think after that I'll just go ahead and merge despite the failing upload, assuming I have the permissions for that. I'll also open a github request with axim to have a look at the coverage problem if nobody has done that yet.

jesperhodge reviewed (approved):
I'm not sure @cablaa77 is available today, so I'm approving this. I don't expect any pushback on the UX and since this is behind a feature flag it's fine to merge.

jesperhodge commented:
@mariajgrimaldi sorry this is all a lot of back and forth and taking a while. You can see I opened a request with axim to look into the CodeCov situation. I plan to just wait for them to get back to me, so if there's any need to merge this sooner than that please just notify me and I can take care of it.

mariajgrimaldi commented:
Thank you for all the help and the review, @jesperhodge! Due to time constraints, we'd like to merge this PR soon. If the coverage check is a blocking issue, could we check the report another way? 

Feel free to let me know if you folks need anything else from us. Thanks again!

jesperhodge commented:
Hi @mariajgrimaldi , if you rebase or merge master into your branch, it should fix the pipeline.

mariajgrimaldi commented:
@jesperhodge: done! Thank you so much for all the help :)

openedx-webhooks commented:
@johnvente 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/frontend-app-course-authoring&entry.867055334=https://github.com/openedx/frontend-app-course-authoring/pull/846&entry.1484655318=https://github.com/johnvente&entry.752974735=2024-02-20+01:40&entry.1917517419=2024-04-11+20:25&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
